_user/serverdemo/main.c

```diff
@@ -0,0 +1,191 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Server demo
+ *
+ * Example of Phoenix server application
+ *
+ * Copyright 2024 Phoenix Systems
+ * Author: Aleksander Kaminski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <sys/types.h>
+
+/* Message handling */
+#include <sys/msg.h>
+
+/* create_dev() */
+#include <posix/utils.h>
+
+
+static int server_handleOpen(oid_t *oid)
+{
+	/* Just allow open() on our interface.
+	 * We need to handle this to allow open()
+	 * to work with our server. */
+	(void)oid;
+	return 0;
+}
+
+
+static int server_handleClose(oid_t *oid)
+{
+	/* Just allow close() on our interface.
+	 * We need to handle this to allow close()
+	 * to work with our server. */
+	(void)oid;
+	return 0;
+}
+
+
+static ssize_t server_handleRead(oid_t *oid, void *data, size_t len, off_t offset)
+{
+	/* This is where we handle read request (i.e. user is reading from the server). */
+	printf("serverdemo: Read from oid %u:%u of %zu bytes @offset %lld\n",
+		(unsigned)oid->port, (unsigned)oid->id, len, offset);
+
+	/* Put something into the requester buffer. */
+	memset(data, 'x', len);
+
+	/* Actual read length or error (negative value). */
+	return (ssize_t)len;
+}
+
+
+static ssize_t server_handleWrite(oid_t *oid, const void *data, size_t len, off_t offset)
+{
+	/* This is where we handle write request (i.e. user is writing to the server). */
+	printf("serverdemo: Write to oid %u:%u of %zu bytes @offset %lld\n",
+		(unsigned)oid->port, (unsigned)oid->id, len, offset);
+
+	/* Print out received data. */
+	printf("Data:");
+	for (size_t i = 0; i < len; i += 16) {
+		printf("\n%08x: ", (unsigned int)i);
+		for (size_t j = 0; (j < 16) && ((i + j) < len); ++j) {
+			printf("%02x ", ((unsigned char *)data)[i + j]);
+		}
+	}
+	printf("\n");
+
+	/* Actual write length or error (negative value). */
+	return (ssize_t)len;
+}
+
+
+__attribute__((noreturn)) static void server_msgLoop(oid_t *oid)
+{
+	/* This is a message structure. It is filled-in by kernel during
+	 * msgRecv() syscall. It contains all data necessary to handle
+	 * the request. */
+	msg_t msg;
+
+	/* This is a message respond ID. It allows the kernel to disting
+	 * between messages passed to the server when the server is responding.
+	 * This is needed, because server might be handling more than one
+	 * message in parallel in any given time, so the kernel needs an
+	 * information, to which user is the server responding to. */
```

Comment:
```suggestion
	 * It is needed, because the server might be handling more than one
	 * message in parallel at any given time, so the kernel needs an
	 * information to which user the server is responding. */
```